src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -0,0 +1,151 @@
+                    list.Add(current.Header);
+                    current = current.Previous;
+                    count--;
+                }
+
+                list.Reverse();
+
+                return list;
+            }
+
+            public bool ConnectedHeadersIsEmpty(ConnectedHeaders connectedHeaders)
+            {
+                Assert.NotNull(connectedHeaders);
+
+                return connectedHeaders.Consumed == null 
+                       && connectedHeaders.DownloadTo == null 
+                       && connectedHeaders.DownloadFrom == null;
+            }
+        }
+
+        [Fact]
+        public void ConnectHeaders_HeadersCantConnect_ShouldFail()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+
+            Assert.Throws<ConnectHeaderException>(() => chainedHeaderTree.ConnectNewHeaders(1, new List<BlockHeader>(new [] { testContext.Network.GetGenesis().Header})));
+        }
+
+        [Fact]
+        public void ConnectHeaders_NoNewHeadersToConnect_ShouldReturnNothingToDownload()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+
+            var chainTip = testContext.ExtendAChain(10);
+            chainedHeaderTree.Initialize(chainTip);
+
+            var listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 4);
+
+            var connectedHeaders = chainedHeaderTree.ConnectNewHeaders(1, listOfExistingHeaders);
+
+            Assert.True(testContext.ConnectedHeadersIsEmpty(connectedHeaders));
+        }
+
+        [Fact]
+        public void ConnectHeaders_HeadersFromTwoPeers_ShouldCreateTwoPeerTips()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+
+            var chainTip = testContext.ExtendAChain(10);
+            chainedHeaderTree.Initialize(chainTip);
+
+            var listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 4);
+
+            var connectedHeaders1 = chainedHeaderTree.ConnectNewHeaders(1, listOfExistingHeaders);
+            var connectedHeaders2 = chainedHeaderTree.ConnectNewHeaders(2, listOfExistingHeaders);
+
+            Assert.Single(chainedHeaderTree.GetPeerIdsByTipHash);
+            Assert.Equal(10, chainedHeaderTree.GetChainedHeadersByHash.Count);
+
+            Assert.Equal(3, chainedHeaderTree.GetPeerIdsByTipHash.First().Value.Count);
+
+            Assert.Equal(ChainedHeaderTree.LocalPeerId, chainedHeaderTree.GetPeerIdsByTipHash.First().Value.ElementAt(0));
+            Assert.Equal(1, chainedHeaderTree.GetPeerIdsByTipHash.First().Value.ElementAt(1));
+            Assert.Equal(2, chainedHeaderTree.GetPeerIdsByTipHash.First().Value.ElementAt(2));
+
+            Assert.True(testContext.ConnectedHeadersIsEmpty(connectedHeaders1));
```

Comment:
I am confused by this connectedHeadersIsEmpty method call

Reply:
If you look at the top you'll see that all internal property are null, this means nothing is requested to be downloaded.

Hmm yeah perhaps thats a better name `EnsureNothingNeedsToDownload` or so?